src/tree_data.c

```diff
@@ -2336,6 +2339,987 @@ lyd_merge(struct lyd_node **target, const struct lyd_node *source, int options)
+            if (oper) {
+                ret = lyd_diff_add(iter_second, oper, orig_default, orig_value, key, value, orig_key, diff);
+
+                free(orig_value);
+                free(key);
+                free(value);
+                free(orig_key);
+                LY_CHECK_GOTO(ret, cleanup);
+            }
+        } else if (!match_first) {
+            /* get all the attributes */
+            LY_CHECK_GOTO(lyd_diff_attrs(match_first, iter_second, options, &oper, &orig_default, &orig_value), cleanup);
+
+            /* there must be changes, it is created */
+            assert(oper && (oper[0] == 'c'));
+            ret = lyd_diff_add(iter_second, oper, orig_default, orig_value, NULL, NULL, NULL, diff);
+
+            free(orig_value);
+            LY_CHECK_GOTO(ret, cleanup);
+        } /* else was handled */
+
+        if (nosiblings) {
+            break;
+        }
+    }
+
+cleanup:
+    LY_ARRAY_FOR(userord, u) {
+        LY_ARRAY_FREE(userord[u].inst);
+    }
+    LY_ARRAY_FREE(userord);
+    return ret;
+}
+
+API LY_ERR
+lyd_diff(const struct lyd_node *first, const struct lyd_node *second, int options, struct lyd_node **diff)
+{
+    const struct ly_ctx *ctx;
+
+    LY_CHECK_ARG_RET(NULL, diff, LY_EINVAL);
+
+    if (first) {
+        ctx = LYD_NODE_CTX(first);
+    } else if (second) {
+        ctx = LYD_NODE_CTX(second);
+    } else {
+        ctx = NULL;
+    }
+
+    if (first && second && (lysc_data_parent(first->schema) != lysc_data_parent(second->schema))) {
+        LOGERR(ctx, LY_EINVAL, "Invalid arguments - cannot create diff for unrelated data (%s()).", __func__);
+        return LY_EINVAL;
+    }
+
+    *diff = NULL;
+
+    return lyd_diff_siblings_r(first, second, options, diff);
+}
+
+/**
+ * @brief Find a matching node in data tree for a diff node.
+ *
+ * @param[in] first_node First sibling in the data tree.
+ * @param[in] diff_node Diff node to match.
+ * @param[out] match_p Matching node.
+ * @return LY_ERR value.
+ */
+static LY_ERR
+lyd_diff_find_node(const struct lyd_node *first_node, const struct lyd_node *diff_node, struct lyd_node **match_p)
+{
+    if (diff_node->schema->nodetype & (LYS_LIST | LYS_LEAFLIST)) {
+        /* try to find the exact instance */
+        lyd_find_sibling_first(first_node, diff_node, match_p);
+    } else {
+        /* try to simply find the node, there cannot be more instances */
+        lyd_find_sibling_val(first_node, diff_node->schema, NULL, 0, match_p);
+    }
+    LY_CHECK_ERR_RET(!*match_p, LOGINT(LYD_NODE_CTX(diff_node)), LY_EINT);
+
+    return LY_SUCCESS;
+}
+
+/**
+ * @brief Learn operation of a diff node.
+ *
+ * @param[in] diff_node Diff node.
+ * @param[out] op Operation.
+ * @param[out] key_or_value Optional list instance keys predicate or leaf-list value for move operation.
+ * @return LY_ERR value.
+ */
+static LY_ERR
+lyd_diff_get_op(const struct lyd_node *diff_node, const char **op, const char **key_or_value)
```

Comment:
I would prefer to have the `op` as an enum - checking it in ifs against some explicit enum value with a clear meaning would be much better than deciding the following steps based on the first character of the operation name. In case of adding some new operation sometime in the future, this approach based on the different leading characters can simply invalidate.

Reply:
Changed.